pygmt/tests/test_geopandas.py

```diff
@@ -131,3 +132,44 @@ def test_geopandas_plot3d_non_default_circle():
         style="c0.2c",
     )
     return fig
+
+
+@pytest.mark.parametrize("dtype", ["int32", "int64", pd.Int32Dtype(), pd.Int64Dtype()])
+@pytest.mark.mpl_image_compare(filename="test_geopandas_plot_int_dtypes.png")
+def test_geopandas_plot_int_dtypes(dtype):
+    """
+    Check that plotting a geopandas GeoDataFrame with integer columns works,
+    including int32 and int64 (non-nullable), Int32 and Int64 (nullable).
+
+    This is a regression test for
+    https://github.com/GenericMappingTools/pygmt/issues/2497
+    """
+    # Read shapefile in geopandas.GeoDataFrame
+    shapefile = which(
+        fname="@RidgeTest.shp @RidgeTest.shx @RidgeTest.dbf @RidgeTest.prj",
+        download="c",
+    )
+    gdf = gpd.read_file(shapefile[0])
+
+    # Reproject geometry and change dtype of NPOINTS column
+    gdf["geometry"] = (
+        gdf.to_crs(crs="EPSG:3857")
+        .buffer(distance=100000)
+        .to_crs(crs="OGC:CRS84")  # convert to lon/lat to prevent @null in PROJ CRS
+    )
+    gdf["NPOINTS"] = gdf.NPOINTS.astype(dtype=dtype)
+
+    # Plot figure with three polygons colored based on NPOINTS value
+    fig = Figure()
+    makecpt(cmap="lajolla", series=[10, 60, 10], continuous=True)
```

Comment:
We might need to check what happens when a big int64 number $2^{32} > x > 2^{64}$ is passed in. Don't want to accidentally convert and plot data incorrectly!

Reply:
@seisman, could you add a test for a large int64 number too? Can use a simple `info` just to check the cast to int32 behaviour, want tomake sure it doesn't truncate int64 numbers to int32.

Reply:
Modified the existing test with large numbers:
```
import geopandas as gpd
from pygmt import which, info

shapefile = which(                                                          
    fname="@RidgeTest.shp @RidgeTest.shx @RidgeTest.dbf @RidgeTest.prj",       
    download="c",                                                           
)                                                                           
gdf = gpd.read_file(shapefile[0])

gdf["geometry"] = (                                                         
    gdf.to_crs(crs="EPSG:3857")                                             
    .buffer(distance=100000)                                                
    .to_crs(crs="OGC:CRS84")  # convert to lon/lat to prevent @null in PROJ CRS
)

gdf["NPOINTS"] *= 1.0e10
gdf["NPOINTS"] = gdf.NPOINTS.astype(dtype="int64")
info(gdf)
```
It raises an OverflowError error:
```
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
Cell In[12], line 1
----> 1 info(gdf)

File ~/OSS/gmt/pygmt/pygmt/helpers/decorators.py:598, in use_alias.<locals>.alias_decorator.<locals>.new_module(*args, **kwargs)
    591     msg = (
    592         "Parameters 'Y' and 'yshift' are deprecated since v0.8.0. "
    593         "and will be removed in v0.12.0. "
    594         "Use Figure.shift_origin(yshift=...) instead."
    595     )
    596     warnings.warn(msg, category=SyntaxWarning, stacklevel=2)
--> 598 return module_func(*args, **kwargs)

File ~/OSS/gmt/pygmt/pygmt/helpers/decorators.py:738, in kwargs_to_strings.<locals>.converter.<locals>.new_module(*args, **kwargs)
    736             kwargs[arg] = separators[fmt].join(f"{item}" for item in value)
    737 # Execute the original function and return its output
--> 738 return module_func(*args, **kwargs)

File ~/OSS/gmt/pygmt/pygmt/src/info.py:85, in info(data, **kwargs)
     83 file_context = lib.virtualfile_from_data(check_kind="vector", data=data)
     84 with GMTTempFile() as tmpfile:
---> 85     with file_context as fname:
     86         lib.call_module(
     87             module="info",
     88             args=build_arg_string(kwargs, infile=fname, outfile=tmpfile.name),
     89         )
     90     result = tmpfile.read()

File ~/opt/miniconda/lib/python3.10/contextlib.py:135, in _GeneratorContextManager.__enter__(self)
    133 del self.args, self.kwds, self.func
    134 try:
--> 135     return next(self.gen)
    136 except StopIteration:
    137     raise RuntimeError("generator didn't yield") from None

File ~/OSS/gmt/pygmt/pygmt/helpers/tempfile.py:147, in tempfile_from_geojson(geojson)
    145     ogrgmt_kwargs["schema"] = schema
    146     # Using geopandas.to_file to directly export to OGR_GMT format
--> 147     geojson.to_file(**ogrgmt_kwargs)
    148 except AttributeError:
    149     # Other 'geo' formats which implement __geo_interface__
    150     import json

File ~/opt/miniconda/lib/python3.10/site-packages/geopandas/geodataframe.py:1263, in GeoDataFrame.to_file(self, filename, driver, schema, index, **kwargs)
   1172 """Write the ``GeoDataFrame`` to a file.
   1173 
   1174 By default, an ESRI shapefile is written, but any OGR data source
   (...)
   1259 
   1260 """
   1261 from geopandas.io.file import _to_file
-> 1263 _to_file(self, filename, driver, schema, index, **kwargs)

File ~/opt/miniconda/lib/python3.10/site-packages/geopandas/io/file.py:572, in _to_file(df, filename, driver, schema, index, mode, crs, engine, **kwargs)
    569     raise ValueError(f"'mode' should be one of 'w' or 'a', got '{mode}' instead")
    571 if engine == "fiona":
--> 572     _to_file_fiona(df, filename, driver, schema, crs, mode, **kwargs)
    573 elif engine == "pyogrio":
    574     _to_file_pyogrio(df, filename, driver, schema, crs, mode, **kwargs)

File ~/opt/miniconda/lib/python3.10/site-packages/geopandas/io/file.py:601, in _to_file_fiona(df, filename, driver, schema, crs, mode, **kwargs)
    597     crs_wkt = crs.to_wkt("WKT1_GDAL")
    598 with fiona.open(
    599     filename, mode=mode, driver=driver, crs_wkt=crs_wkt, schema=schema, **kwargs
    600 ) as colxn:
--> 601     colxn.writerecords(df.iterfeatures())

File ~/opt/miniconda/lib/python3.10/site-packages/fiona/collection.py:558, in Collection.writerecords(self, records)
    556 if self.mode not in ("a", "w"):
    557     raise OSError("collection not open for writing")
--> 558 self.session.writerecs(records, self)
    559 self._len = self.session.get_length()
    560 self._bounds = None

File fiona/ogrext.pyx:1409, in fiona.ogrext.WritingSession.writerecs()

File fiona/ogrext.pyx:441, in fiona.ogrext.OGRFeatureBuilder.build()

OverflowError: value too large to convert to int
```

Reply:
Ok, good to know that it errors. We should probably cast int64 numbers to float64 or string dtype, xref https://github.com/geopandas/geopandas/pull/2950#issuecomment-1620900299 and https://github.com/OSGeo/gdal/blob/6e9103bd5acb1ff8da305c4e77fa30335ffa3b70/ogr/ogrsf_frmts/gmt/ogrgmtlayer.cpp#L754-L773. We can do this in a follow-up PR since this one is getting a bit long.